`E0601: used-before-assignment` false positive for import under `TYPE_CHECKING` check

### Bug description

```python
# pylint: disable=missing-module-docstring, missing-function-docstring

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from typing_extensions import assert_type
else:

    def assert_type(val, _):
        return val

ANSWER = 42
assert_type(ANSWER, int)
```

### Configuration

_No response_

### Command used

```shell
pylint a.py
```


### Pylint output

```shell
************* Module a
a.py:13:0: E0601: Using variable 'assert_type' before assignment (used-before-assignment)

------------------------------------------------------------------
Your code has been rated at 2.86/10 (previous run: 2.86/10, +0.00)
```


### Expected behavior

This is a false positive, as `assert_type` is defined in both branches of the `if`.

### Pylint version

```shell
pylint 2.15.0
astroid 2.12.4
Python 3.10.6 (main, Aug  3 2022, 17:39:45) [GCC 12.1.1 20220730]
```


### OS / Environment

Archlinux

### Additional dependencies

astroid==2.12.4
dill==0.3.5.1
isort==5.10.1
lazy-object-proxy==1.7.1
mccabe==0.7.0
platformdirs==2.5.2
pylint==2.15.0
tomli==2.0.1
tomlkit==0.11.4
wrapt==1.14.1

The issue is still here if the import is from stdlib instead of an external module:

```python
# pylint: disable=missing-module-docstring, missing-function-docstring

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections import defaultdict
else:
    def defaultdict():
        return {}

defaultdict()
```

```
************* Module a
c.py:11:6: E0601: Using variable 'defaultdict' before assignment (used-before-assignment)

------------------------------------------------------------------
Your code has been rated at 1.67/10 (previous run: 1.67/10, +0.00)
```

--- a/doc/whatsnew/fragments/7368.false_positive
+++ b/doc/whatsnew/fragments/7368.false_positive
@@ -1,0 +1,3 @@
+Fix ``used-before-assignment`` for functions/classes defined in type checking guard.
+
+Closes #7368

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -1841,7 +1841,6 @@ class VariablesChecker(BaseChecker):
         base_scope_type,
         is_recursive_klass,
     ) -> tuple[bool, bool, bool]:
-        # pylint: disable=too-many-nested-blocks
         maybe_before_assign = True
         annotation_return = False
         use_outer_definition = False
@@ -2014,8 +2013,13 @@ class VariablesChecker(BaseChecker):
                                 for target in definition.targets
                                 if isinstance(target, nodes.AssignName)
                             )
-                            if defined_in_or_else:
-                                break
+                        elif isinstance(
+                            definition, (nodes.ClassDef, nodes.FunctionDef)
+                        ):
+                            defined_in_or_else = definition.name == node.name
+
+                        if defined_in_or_else:
+                            break
 
                     if not used_in_branch and not defined_in_or_else:
                         maybe_before_assign = True

--- a/tests/functional/u/undefined/undefined_variable.py
+++ b/tests/functional/u/undefined/undefined_variable.py
@@ -267,15 +267,24 @@ if TYPE_CHECKING:
 if TYPE_CHECKING:
     from collections import Counter
     from collections import OrderedDict
+    from collections import defaultdict
+    from collections import UserDict
 else:
     Counter = object
     OrderedDict = object
+    def defaultdict():
+        return {}
+    class UserDict(dict):
+        pass
 
 
 def tick(counter: Counter, name: str, dictionary: OrderedDict) -> OrderedDict:
     counter[name] += 1
     return dictionary
 
+defaultdict()
+
+UserDict()
 
 # pylint: disable=unused-argument
 def not_using_loop_variable_accordingly(iterator):

--- a/tests/functional/u/undefined/undefined_variable.txt
+++ b/tests/functional/u/undefined/undefined_variable.txt
@@ -28,12 +28,12 @@ undefined-variable:171:4:171:13::Undefined variable 'unicode_3':UNDEFINED
 undefined-variable:226:25:226:37:LambdaClass4.<lambda>:Undefined variable 'LambdaClass4':UNDEFINED
 undefined-variable:234:25:234:37:LambdaClass5.<lambda>:Undefined variable 'LambdaClass5':UNDEFINED
 used-before-assignment:255:26:255:34:func_should_fail:Using variable 'datetime' before assignment:HIGH
-undefined-variable:282:18:282:24:not_using_loop_variable_accordingly:Undefined variable 'iteree':UNDEFINED
-undefined-variable:299:27:299:28:undefined_annotation:Undefined variable 'x':UNDEFINED
-used-before-assignment:300:7:300:8:undefined_annotation:Using variable 'x' before assignment:HIGH
-undefined-variable:330:11:330:12:decorated3:Undefined variable 'x':UNDEFINED
-undefined-variable:335:19:335:20:decorated4:Undefined variable 'y':UNDEFINED
-undefined-variable:356:10:356:20:global_var_mixed_assignment:Undefined variable 'GLOBAL_VAR':HIGH
-undefined-variable:368:19:368:44:RepeatedReturnAnnotations.x:Undefined variable 'RepeatedReturnAnnotations':UNDEFINED
-undefined-variable:370:19:370:44:RepeatedReturnAnnotations.y:Undefined variable 'RepeatedReturnAnnotations':UNDEFINED
-undefined-variable:372:19:372:44:RepeatedReturnAnnotations.z:Undefined variable 'RepeatedReturnAnnotations':UNDEFINED
+undefined-variable:291:18:291:24:not_using_loop_variable_accordingly:Undefined variable 'iteree':UNDEFINED
+undefined-variable:308:27:308:28:undefined_annotation:Undefined variable 'x':UNDEFINED
+used-before-assignment:309:7:309:8:undefined_annotation:Using variable 'x' before assignment:HIGH
+undefined-variable:339:11:339:12:decorated3:Undefined variable 'x':UNDEFINED
+undefined-variable:344:19:344:20:decorated4:Undefined variable 'y':UNDEFINED
+undefined-variable:365:10:365:20:global_var_mixed_assignment:Undefined variable 'GLOBAL_VAR':HIGH
+undefined-variable:377:19:377:44:RepeatedReturnAnnotations.x:Undefined variable 'RepeatedReturnAnnotations':UNDEFINED
+undefined-variable:379:19:379:44:RepeatedReturnAnnotations.y:Undefined variable 'RepeatedReturnAnnotations':UNDEFINED
+undefined-variable:381:19:381:44:RepeatedReturnAnnotations.z:Undefined variable 'RepeatedReturnAnnotations':UNDEFINED

